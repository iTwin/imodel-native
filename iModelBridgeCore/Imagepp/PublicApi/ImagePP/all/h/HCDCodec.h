//:>--------------------------------------------------------------------------------------+
//:>
//:>     $Source: PublicApi/ImagePP/all/h/HCDCodec.h $
//:>
//:>  $Copyright: (c) 2015 Bentley Systems, Incorporated. All rights reserved. $
//:>
//:>+--------------------------------------------------------------------------------------
//-----------------------------------------------------------------------------
// Class : HCDCodec
//-----------------------------------------------------------------------------
// General class for codecs.
//-----------------------------------------------------------------------------

#pragma once

#include <Imagepp/all/h/HFCExclusiveKey.h>

BEGIN_IMAGEPP_NAMESPACE

class HCDPacketRLE;

// Binary optimization: Speed up read/write when codec support direct decompression/compression to/from RLE format.
class HNOVTABLEINIT HCDCodecRLEInterface
    {
public:
    virtual void     DecompressSubsetToRLE(const void* pi_pInData, size_t pi_InDataSize, HFCPtr<HCDPacketRLE>& pio_rpRLEPacket) = 0;
    virtual size_t   CompressSubsetFromRLE(HFCPtr<HCDPacketRLE> const& pi_rpPacketRLE, void* po_pOutBuffer, size_t po_OutBufferSize) = 0;
    };

class HNOVTABLEINIT HCDCodec : public HFCShareableObject<HCDCodec>,
    public HFCExclusiveKey     // to have access to claim/release instead of including m_Key and adding the methods
    {
    HDECLARE_BASECLASS_ID(HCDCodecId_Base)

public:

    // codec states enumeration
    enum State
        {
        STATE_NONE,
        STATE_COMPRESS,
        STATE_DECOMPRESS
        };

    // destructor
    virtual         ~HCDCodec();

    // compression
    virtual size_t          CompressSubset(const void* pi_pInData,
                                           size_t pi_InDataSize,
                                           void* po_pOutBuffer,
                                           size_t pi_OutBufferSize) = 0;

    virtual size_t          DecompressSubset(const void* pi_pInData,
                                             size_t pi_InDataSize,
                                             void* po_pOutBuffer,
                                             size_t pi_OutBufferSize) = 0;

    State                   GetCurrentState() const;
    virtual void            Reset();

    virtual HCDCodec*       Clone() const = 0;

    virtual size_t          GetDataSize() const = 0;

    // subset
    virtual size_t          GetMinimumSubsetSize() const = 0;
    virtual size_t          GetSubsetMaxCompressedSize() const = 0;
    virtual void            SetSubsetSize(size_t pi_Size) = 0;

    // others
    virtual double          GetEstimatedCompressionRatio() const;
    virtual bool            HasRandomAccess() const;
    virtual bool            IsCodecImage() const;
    virtual bool            IsCodecVector() const;
    virtual const WString&  GetName() const;

    virtual HCDCodecRLEInterface* 
                            GetRLEInterface();

protected:

    // constructors
    HCDCodec(const WString& pi_rCodecName);
    HCDCodec(const HCDCodec& pi_rObj);


    // compression
    void                    SetCurrentState(State pi_State);

    WString         m_CodecName;

private:

    State          m_CurrentState;
    };

END_IMAGEPP_NAMESPACE