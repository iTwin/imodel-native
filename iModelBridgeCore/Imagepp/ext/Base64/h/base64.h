#ifndef _base64_h_
#define _base64_h_

#include <iostream>

namespace b64
{
        bool encode(std::istream& input, std::ostream& output);
        bool decode(std::istream& input, std::ostream& output);
};

#endif
