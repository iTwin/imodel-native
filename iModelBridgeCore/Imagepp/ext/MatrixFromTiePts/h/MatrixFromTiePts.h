/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See COPYRIGHT.md in the repository root for full copyright notice.
*--------------------------------------------------------------------------------------------*/

#pragma once


#define TRANSFO_NONE     -1 /* No transformation model */
#define TRANSFO_TRAN_2D   0 /* Translation en x et y */
#define TRANSFO_SIMI_2D   1 /* Translation x,y + Rotation + Echelle */
#define TRANSFO_POLY_2D_1 2 /* Polynomiale deg 1 */
#define TRANSFO_AFFI_2D   3 /* Tran. x,y + Rotation + Echelle + Axes */
#define TRANSFO_PROJ_2D   4 /* Projection d'un plan sur un autre */
#define TRANSFO_POLY_2D_2 5 /* Polynomiale deg 2 */
#define TRANSFO_POLY_2D_3 6 /* Polynomiale deg 3 */
#define TRANSFO_POLY_2D_4 7 /* Polynomiale deg 4 */ /* A tester ??? */
#define TRANSFO_POLY_2D_5 8 /* Polynomiale deg 5 */ /* A tester ??? */
#define TRANSFO_MAT_2D    9 /* Matrice de transformation */
#define TRANSFO_HELM_2D   10
#define TRANSFO_TPS_2D    11
#define TRANSFO_MAX       12

#define TRF_ERROR_MAX_ITERATIONS    -1
#define TRF_ERROR_NO_MEMORY         -2
#define TRF_ERROR_BAD_NUMBER_POINTS -3
#define TRF_ERROR_MATRIX            -4


long GetTransfoMatrixFromScaleAndTiePts (double          po_pMatrix[4][4],
                                         unsigned short  pi_NbVal_GeoTiePoint,
                                         double*         pi_pVal_GeoTiePoint,
                                         unsigned short  pi_NbVal_GeoPixelScale,
                                         double*         pi_pVal_GeoPixelScale);

long GetProjectiveTransfoMatrixFromScaleAndTiePts(double         po_pMatrix[4][4],
                                                  unsigned short pi_NbVal_GeoTiePoint,
                                                  double*        pi_pVal_GeoTiePoint);

long GetAffineTransfoMatrixFromScaleAndTiePts(double         po_pMatrix[4][4],
                                              unsigned short pi_NbVal_GeoTiePoint,
                                              double*        pi_pVal_GeoTiePoint);

long GetTransfoMatrixFromTiePts(double           po_pMatrix[4][4],
                                unsigned short   pi_NbVal_GeoTiePoint,
                                const double*    pi_pVal_GeoTiePoint,
                                long             pi_ModelType);

