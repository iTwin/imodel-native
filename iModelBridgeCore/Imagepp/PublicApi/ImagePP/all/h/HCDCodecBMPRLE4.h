//:>--------------------------------------------------------------------------------------+
//:>
//:>
//:>  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
//:>
//:>+--------------------------------------------------------------------------------------
// Class : HCDCodecBMPRLE8
//-----------------------------------------------------------------------------
// RLE8 codec lass.
//-----------------------------------------------------------------------------
#pragma once

#include "HCDCodecImage.h"

BEGIN_IMAGEPP_NAMESPACE

class HCDCodecBMPRLE4 : public HCDCodecImage
    {
    HDECLARE_CLASS_ID(HCDCodecId_BMPRLE4, HCDCodecImage)

public:

    IMAGEPP_EXPORT                 HCDCodecBMPRLE4();

    HCDCodecBMPRLE4(size_t pi_Width,
                    size_t pi_Height);

    HCDCodecBMPRLE4(const HCDCodecBMPRLE4& pi_rObj);

    ~HCDCodecBMPRLE4();

    size_t          CompressSubset(const void* pi_pInData,
                                   size_t     pi_InDataSize,
                                   void*      po_pOutBuffer,
                                   size_t     pi_OutBufferSize) override;

    size_t          DecompressSubset(const void* pi_pInData,
                                     size_t       pi_InDataSize,
                                     void*        po_pOutBuffer,
                                     size_t       pi_OutBufferSize) override;

    bool           HasLineAccess() const override;

    virtual HCDCodec* Clone() const override;

    bool           IsBitsPerPixelSupported(size_t pi_Bits) const override;

    size_t          GetSubsetMaxCompressedSize() const override;

protected:

private:
    };

END_IMAGEPP_NAMESPACE
