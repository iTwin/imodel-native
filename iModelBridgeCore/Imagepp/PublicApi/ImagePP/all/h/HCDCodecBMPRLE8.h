//:>--------------------------------------------------------------------------------------+
//:>
//:>     $Source: PublicApi/ImagePP/all/h/HCDCodecBMPRLE8.h $
//:>
//:>  $Copyright: (c) 2015 Bentley Systems, Incorporated. All rights reserved. $
//:>
//:>+--------------------------------------------------------------------------------------
// Class : HCDCodecBMPRLE8
//-----------------------------------------------------------------------------
// RLE8 codec lass.
//-----------------------------------------------------------------------------
#pragma once

#include "HCDCodecImage.h"

BEGIN_IMAGEPP_NAMESPACE

class HCDCodecBMPRLE8 : public HCDCodecImage
    {
    HDECLARE_CLASS_ID(HCDCodecId_BMPRLE8, HCDCodecImage)

public:

    IMAGEPP_EXPORT                 HCDCodecBMPRLE8();

    HCDCodecBMPRLE8(  size_t pi_Width,
                      size_t pi_Height);

    HCDCodecBMPRLE8(const HCDCodecBMPRLE8& pi_rObj);

    ~HCDCodecBMPRLE8();

    size_t          CompressSubset(const void* pi_pInData,
                                   size_t pi_InDataSize,
                                   void* po_pOutBuffer,
                                   size_t pi_OutBufferSize);

    size_t          DecompressSubset(const void* pi_pInData,
                                     size_t pi_InDataSize,
                                     void* po_pOutBuffer,
                                     size_t pi_OutBufferSize);

    bool           HasLineAccess() const;

    virtual HCDCodec* Clone() const override;

    bool           IsBitsPerPixelSupported(size_t pi_Bits) const;

    size_t          GetSubsetMaxCompressedSize() const;

protected:

private:
    };

END_IMAGEPP_NAMESPACE