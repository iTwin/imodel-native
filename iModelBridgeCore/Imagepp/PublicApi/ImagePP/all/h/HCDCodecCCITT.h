//:>--------------------------------------------------------------------------------------+
//:>
//:>
//:>  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
//:>
//:>+--------------------------------------------------------------------------------------
//-----------------------------------------------------------------------------
// Class : HCDCodecCCITT
//-----------------------------------------------------------------------------
// CCITT codec lass.
//-----------------------------------------------------------------------------

#pragma once

#include "HCDCodecImage.h"

BEGIN_IMAGEPP_NAMESPACE

class HCDCodecCCITT : public HCDCodecImage
    {
    HDECLARE_CLASS_ID(HCDCodecId_CCITT, HCDCodecImage)

public:

    ~HCDCodecCCITT();

    bool           IsBitsPerPixelSupported(size_t pi_Bits) const override;

    HCDCodecCCITT(const HCDCodecCCITT& pi_rObj);

    size_t          GetSubsetMaxCompressedSize() const override;

    double GetEstimatedCompressionRatio() const override;

    IMAGEPP_EXPORT void     SetPhotometric(uint16_t pi_Photo);
    uint16_t GetPhotometric() const;

    IMAGEPP_EXPORT void     SetBitRevTable(bool pi_Reverse);

protected:
    
    virtual HCDCodec* Clone() const override = 0;

    HCDCodecCCITT();

    HCDCodecCCITT(size_t pi_Width,
                  size_t pi_Height);

    bool           m_bitrevtable;
    uint16_t m_photometric;
    };

END_IMAGEPP_NAMESPACE
