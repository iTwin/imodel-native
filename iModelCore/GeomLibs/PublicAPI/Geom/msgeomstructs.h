/*--------------------------------------------------------------------------------------+
|
|     $Source: PublicAPI/Geom/msgeomstructs.h $
|
|  $Copyright: (c) 2011 Bentley Systems, Incorporated. All rights reserved. $
|
+--------------------------------------------------------------------------------------*/
#pragma once

/*__BENTLEY_INTERNAL_ONLY__*/

#include <Geom/GeomApi.h>


