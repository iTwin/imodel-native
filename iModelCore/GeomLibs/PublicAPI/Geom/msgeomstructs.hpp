/*--------------------------------------------------------------------------------------+
|
|     $Source: PublicAPI/Geom/msgeomstructs.hpp $
|
|  $Copyright: (c) 2009 Bentley Systems, Incorporated. All rights reserved. $
|
+--------------------------------------------------------------------------------------*/
#pragma once
/*__BENTLEY_INTERNAL_ONLY__*/
#include <Geom/GeomApi.h>
