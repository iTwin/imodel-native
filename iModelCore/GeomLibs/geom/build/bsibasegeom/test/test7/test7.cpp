/*--------------------------------------------------------------------------------------+
|
|  Copyright (c) Bentley Systems, Incorporated. All rights reserved.
|
+--------------------------------------------------------------------------------------*/
// test1.cpp : Defines the entry point for the console application.
//

#include <stdio.h>
#include <msgeomstructs.hpp>
#include <msgeomstructs.h>
#include <printfuncs.h>
#include <stdlib.h>

typedef enum
    {
    VREL_Components,
    VREL_UnorientedAngle,
    VREL_OrientedAngle
    } VectorRelationship;

void checkDPlane3d
(
DPlane3d const *pPlaneA,
DPlane3d const *pPlaneB,
char *pDescr,
bool bSameOrigin,
VectorRelationship vrel
)
    {
    char descr[2048];
    if (bSameOrigin)
        {
        sprintf (descr, "DPlane3d.origin.identical::%s", pDescr);
        checkDPoint3d (&pPlaneA->origin, &pPlaneB->origin, descr);
        }
    else
        {
        double dA, dB;
        dA = pPlaneA->evaluate (&pPlaneB->origin);
        dB = pPlaneB->evaluate (&pPlaneA->origin);
        sprintf (descr, "DPlane3d.originB.onPlaneA::%s", pDescr);
        checkDouble (dA, 0.0, descr);
        sprintf (descr, "DPlane3d.originA.onPlaneB::%s", pDescr);
        checkDouble (dB, 0.0, descr);
        }

    if (vrel == VREL_Components)
        {
        sprintf (descr, "DPlane3d.normal.identical::%s", pDescr);
        checkDVec3d (&pPlaneA->normal, &pPlaneB->normal, descr);
        }
    else if (vrel == VREL_OrientedAngle)
        {
        double theta = bsiDVec3d_angleBetweenVectors (&pPlaneA->normal, &pPlaneB->normal);
        sprintf (descr, "DPlane3d.angleBetweenNormals::%s", pDescr);
        checkDouble (theta, 0.0, descr);
        }
    else
        {
        double theta = bsiDVec3d_smallerAngleBetweenUnorientedVectors (&pPlaneA->normal, &pPlaneB->normal);
        sprintf (descr, "DPlane3d.angleBetweenUnorientedNormals::%s", pDescr);
        checkDouble (theta, 0.0, descr);
        }
    }

void test_DPlane3d ()
    {
    DPlane3d planeA, planeB, planeC, planeD;
    DPoint3d origin;
    DVec3d normal;
    origin.init (1,2,3);
    normal.init (0.2, 0.3, 0.4);
    planeA.initFromOriginAndNormal (origin.x, origin.y, origin.z, normal.x, normal.y, normal.z);
    planeB.initFromOriginAndNormal (&origin, &normal);
    checkDPlane3d (&planeA, &planeB, "DPlane3d.initFromOriginAndNormal");
    checkDPlane3d (&planeA, &planeB, "DPlane3d.initFromOriginAndNormal", true, VREL_Components);

    planeB.normalize ();
    checkDPlane3d (&planeA, &planeB, "DPlane3d.normalize", true, VREL_OrientedAngle);
    checkDouble (1.0, planeB.normal.magnitude (), "DPlane3d.normalize");

    DPoint4d hPlaneA;
    planeA.getDPoint4d (&hPlaneA);
    planeC.init (&hPlaneA);
    checkDPlane3d (&planeA, &planeC, "DPlane3d HPlane RT", false, VREL_UnorientedAngle);

    double aa, bb, cc, dd;
    planeA.getCoefficients (&aa, &bb, &cc, &dd);

    planeD.init (aa, bb, cc, dd);
    checkDPlane3d (&planeA, &planeD, "DPlane3d Coffs RT", false, VREL_UnorientedAngle);

    checkFalse (planeD.isZero (), "DPlane3d.isZero");
    planeD.zero ();
    checkTrue  (planeD.isZero (), "DPlane3d.isZero");

    DPoint3d points[3];
    points[0].init (1,2,3);
    points[1].init (4,3,9);
    points[2].init (-2,5,3);
    DPlane3d planeE, planeF;
    planeE.initFrom3Points (&points[0], &points[1], &points[2]);
    planeF.initFromArray   (points, 3);
    checkDPlane3d (&planeE, &planeF, "DPlane3d 3pts, array", false, VREL_UnorientedAngle);

    for (int i = 0; i < 3; i++)
        {
        checkDouble (planeE.evaluate (&points[i]), 0.0, "DPlane3d evaluate (E)");
        checkDouble (planeF.evaluate (&points[i]), 0.0, "DPlane3d evaluate (F)");
        }
    DPoint3d pointQ, pointR;
    double a = 1.23123;
    pointQ.sumOf (&planeE.origin, &planeE.normal, a);
    checkDouble (planeE.evaluate (&pointQ), a * planeE.normal.magnitudeSquared (), "DPlane3d.evaluate");

    pointQ.sumOf (&points[2], &planeE.normal, a);
    planeE.projectPoint (&pointR, &pointQ);
    checkDPoint3d (&pointR, &points[2], "DPlane3d.projectPoint");


#ifdef abc

bool    DPlane3d::productOf
(

const   DTransform3d  *pTransform,
const   DPlane3d      *pSource

);
#endif
    }

int main(int argc, char * argv[])
    {
    initErrorTracking (NULL, argc, argv);
    test_DPlane3d ();
    return getExitStatus();
    }
