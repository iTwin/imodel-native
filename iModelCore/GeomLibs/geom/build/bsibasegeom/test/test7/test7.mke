#--------------------------------------------------------------------------------------
#
#     $Source: geom/build/bsibasegeom/test/test7/test7.mke $
#
#  $Copyright: (c) 2011 Bentley Systems, Incorporated. All rights reserved. $
#
#--------------------------------------------------------------------------------------
TESTAPP_NAME=test7

%include    $(_MakeFilePath)/../testapp.mki
