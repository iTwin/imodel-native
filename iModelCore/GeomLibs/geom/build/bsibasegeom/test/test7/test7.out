 geomlibs test environment -- command line arguments:
 Tail of argv[0] = bsigeomtest7.exe
<OK name="(DPlane3d.origin)DPlane3d.initFromOriginAndNormal"><DPoint3d>1,2,3</DPoint3d></OK>
<OK name="(DPlane3d.normal)DPlane3d.initFromOriginAndNormal"<DVec3d>0.2,0.3,0.4</DVec3d>/>
<OK name="DPlane3d.origin.identical::DPlane3d.initFromOriginAndNormal"><DPoint3d>1,2,3</DPoint3d></OK>
<OK name="DPlane3d.normal.identical::DPlane3d.initFromOriginAndNormal"<DVec3d>0.2,0.3,0.4</DVec3d>/>
<OK name="DPlane3d.origin.identical::DPlane3d.normalize"><DPoint3d>1,2,3</DPoint3d></OK>
<OK name="DPlane3d.angleBetweenNormals::DPlane3d.normalize"><double>5.76244e-017</double></OK>
<OK name="DPlane3d.normalize"><double>1</double></OK>
<OK name="DPlane3d.originB.onPlaneA::DPlane3d HPlane RT"><double>0</double></OK>
<OK name="DPlane3d.originA.onPlaneB::DPlane3d HPlane RT"><double>0</double></OK>
<OK name="DPlane3d.angleBetweenUnorientedNormals::DPlane3d HPlane RT"><double>5.76244e-017</double></OK>
<OK name="DPlane3d.originB.onPlaneA::DPlane3d Coffs RT"><double>0</double></OK>
<OK name="DPlane3d.originA.onPlaneB::DPlane3d Coffs RT"><double>0</double></OK>
<OK name="DPlane3d.angleBetweenUnorientedNormals::DPlane3d Coffs RT"><double>5.76244e-017</double></OK>
<OK name="DPlane3d.isZero"><bool::false/></OK>
<OK name="DPlane3d.isZero"><bool::true/></OK>
<OK name="DPlane3d.originB.onPlaneA::DPlane3d 3pts, array"><double>0</double></OK>
<OK name="DPlane3d.originA.onPlaneB::DPlane3d 3pts, array"><double>0</double></OK>
<OK name="DPlane3d.angleBetweenUnorientedNormals::DPlane3d 3pts, array"><double>0</double></OK>
<OK name="DPlane3d evaluate (E)"><double>0</double></OK>
<OK name="DPlane3d evaluate (F)"><double>0</double></OK>
<OK name="DPlane3d evaluate (E)"><double>0</double></OK>
<OK name="DPlane3d evaluate (F)"><double>0</double></OK>
<OK name="DPlane3d evaluate (E)"><double>0</double></OK>
<OK name="DPlane3d evaluate (F)"><double>0</double></OK>
<OK name="DPlane3d.evaluate"><double>975.134</double></OK>
<OK name="DPlane3d.projectPoint"><DPoint3d>-2,5,3</DPoint3d></OK>
Exit with 0 errors, 26 OK
